### PR TITLE
Use pivot table endpoints to display subtotals

### PR DESCRIPTION
**Pull request by paulrosenzweig**
Part of #13662

This PR puts starts using the dataset query endpoint. The returned data includes the same aggregates broken down by subsets of the original breakouts. That data is displayed as subtotals alongside the raw data.

image

Note: if you're trying this locally there's a bug where heights don't adjust as you add/remove breakouts. Refreshing fixes that.

### CODE REVIEW

File: frontend/src/metabase-lib/lib/Question.js [hunk 1]
```diff
@@ -824,7 +824,7 @@ export default class Question {
       };
 
       return [
-        await CardApi.query(queryParams, {
+        await maybeUsePivotEndpoint(CardApi.query, this.card())(queryParams, {
```
Comment (kdoh) [inline frontend/src/metabase-lib/lib/Question.js:827]: _Nervous laughter_

What's the story behind the "maybe" here?

File: frontend/src/metabase/services.js [hunk 2]
```diff
@@ -28,6 +32,35 @@ export const GTAPApi = {
   attributes: GET("/api/mt/user/attributes"),
 };
 
+export function maybeUsePivotEndpoint(api: APIMethod, card: Card): APIMethod {
```
Comment (kdoh) [inline frontend/src/metabase/services.js:35]: Were there any other approaches you considered to this? It feels a bit messy to have to go through this code path for cards that aren't displayed as pivots.

File: frontend/src/metabase/services.js [hunk 3]
```diff
@@ -28,6 +32,35 @@ export const GTAPApi = {
...
+            breakout.findIndex(b => _.isEqual(b, field_ref)),
+          ),
+      );
+      return api({ ...params, pivot_rows, pivot_cols }, ...rest);
+    };
+  }
+  if (card.display !== "pivot") {
+    return api;
+  }
+  switch (api) {
```
Comment (kdoh) [inline frontend/src/metabase/services.js:54]: Might be worth leaving a note as to why we have to check these.

File: frontend/src/metabase/query_builder/actions.js [hunk 4]
```diff
@@ -833,6 +834,41 @@ export const updateQuestion = (
       run = false;
     }
 
+    // <PIVOT LOGIC>
```
Comment (kdoh) [inline frontend/src/metabase/query_builder/actions.js:837]: 👍 on the comments here

**Review by kdoh — COMMENTED** [review]
I'm a little nervous about the fact that the pivot related needs have to bubble up to the point where we have to wrap API calls this way. Kinda feels like it's tech debt already. Were there any other options or were the alternatives just as ugly?

**Comment by paulrosenzweig** [comment]
@kdoh thanks for the comments. You hit on the main idea in this PR, which I totally failed to call out in the PR message. 🤦 

> Kinda feels like it's tech debt already. Were there any other options or were the alternatives just as ugly?

Agreed. Here's a full [TDD](https://docs.google.com/document/d/1Sb58EgSCXSzv1LY5MyZGeTLumFQeo8bYEmdT558AdR4/edit), but the TL;DR is:

We had three main options. The one we picked isn't the cleanest, but it should get this over the finish line more quickly.
1. Update MBQL
If we add the concept of "breakout sets" to MBQL we can fetch all needed data with the existing query/card endpoints.
   - It's slow and risky to implement MBQL changes, but it's probably the "right" solution.
   - This will let us avoid potential consistency issues. (Cam pointed out we don't get this for free, but we have the option at least if it's in one query.)
   - The frontend will need to update the query when changing to/from a pivot table, but it won't need to swap out endpoints like this PR does.
2. Add another request to fetch subtotals
This was the original plan. We could still fetch the primary data using the existing endpoints but also hit another endpoint to get subtotals.
   - The hypothetical consistency issues are likely worse here since we increase the lag between fetching primary data and subtotals.
   - The frontend needs more logic to coordinate fetching the other endpoint alongside the main query endpoints. The card can be partially loaded.
3. Use a different endpoint for all pivot data
This option is implemented in this PR. We use new endpoints for pivot tables that completely replace the existing endpoints.
    - This requires the hacky `maybeUsePivotEndpoint` wrappers.
    - This is the medium option for any consistency issues. All data is grouped in one API request, but (currently) separate requests are issued against the db.

To summarize: We started at (2), ended up doing (3), and want to do (1) eventually.

**Comment by kdoh** [comment]
Got it, that makes more sense now.

File: frontend/src/metabase/services.js (hunk 3 above)
Comment (paulrosenzweig) [reply frontend/src/metabase/services.js:58]: added!
